dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb

```diff
@@ -20,12 +20,16 @@ def update
       end
 
       status = params[:status]
+      previous_status = @application.status
       if status
         @application.status = status
       end
 
       if @application.save
         render json: @application, status: :ok
+
+        # send confirmation email only if user is submitting their application for the first time
+        on_successful_create if previous_status == 'incomplete' && status == 'unreviewed'
```

Comment:
Yeah, good call––I'll investigate this further. It looks like we are defining `on_successful_create` in both application controllers (it's in `FormsController` asking for an override) and application models. Since it's in FormsController, which is inherited by all applications, and that's calling `on_successful_create` when an application is created, we'd need to make sure whatever is in the controller moves to the model with a callback.

Are you thinking `on_successful_create` could be called from the model with an `after_create` callback? I could see that working ... although in this particular case (teacher apps that can be saved), using a callback would be hard. We'd have to use an `after_save` callback *and* check the previous status, maybe by parsing the ugly `status_timestamp_change_log` field.

Thoughts?

Reply:
> Yeah, good call––I'll investigate this further. It looks like we are defining on_successful_create in both application controllers (it's in FormsController asking for an override) and application models. Since it's in FormsController, which is inherited by all applications, and that's calling on_successful_create when an application is created, we'd need to make sure whatever is in the controller moves to the model with a callback.

good point, I missed this chain of inheritance. it looks like we are performing this check only on `if form.save`, which makes me think you could potentially scrap all of the `on_successful_create` method definitions in the controllers, and just use an `after_save` callback instead.

> Are you thinking on_successful_create could be called from the model with an after_create callback? I could see that working ... although in this particular case (teacher apps that can be saved), using a callback would be hard. We'd have to use an after_save callback and check the previous status, maybe by parsing the ugly status_timestamp_change_log field.

I'm not sure if this is the only problem, but this particular part is where the `_was` trick would come in handy -- if my understanding is correct, you wouldn't have to look at the change log, you could just look at `status_was`.

If this all adds up to a lot more work than the original fix, then it's fine to pass on this for now, but based on the issues raised so far it still seems to me like it might just work.

Reply:
We're using `current_user` from the controller to log who changed the application status ... does that seem feasible to do from the model? I think `user` in the model is always the applicant

Reply:
Maybe better is to do this directly as in 
https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/api/v1/pd/applications_controller.rb#L181

and then remove the extraneous `on_successful_create` methods ...